### ISSUE TITLE
Release 1.0

### ISSUE THREAD

**Issue by juanibiapina** [lostisland/faraday#746]
Why not release version 1.0 now? It's already used in production by many people all over the world. This might be the single best, most featured client out there, despite sometimes being slightly annoying to use.

My suggestion is to just bump to 1.0 now. I think semantic versioning says somewhere that if it's used in production, it's probably 1.0 already. From there you can move forward with semantic versioning and users don't need to stress over what's going to happen on the next gem update.

**Comment by iMacTia** [comment]
Hi @juanibiapina,

I totally understand what you mean and I personally agree with you as well!
Unfortunately, I believe Faraday missed the opportunity to go "1.0" with version 0.9.2
That version was very stable and used in production by thousands of applications!
However, that never happened and everyone has been used to 0.9.x for too long now.

When I started maintaining Faraday last year, I wanted to move into v1.0 and take advantage of it by doing some backwards incompatible change, but before that I wanted to go through all the issues/PR that were open over time and try to sort them out.
Releasing a 1.0 version at that point, might have turned people into thinking that the new release was backwards incompatible as well, but that wasn't the case, so I kept the 0.x version and kept working on that.

Now I'm pretty happy with what we've got, there are only a bunch of open issues/PRs and the remaining changes are difficult to implement maintaining backwards compatibility.
In short, this is where we really start working on v1.0 😃
Jumping to it now would just create confusion at this point, plus when we'll merge the remaining changes we'll need to go to v2.0 as they will brake backwards compatibility again 😅

**Comment by juanibiapina** [comment]
The opportunity was missed them, I agree, but also is "being missed" every day ever since. It's not too late to fix it. The reason you give in the end, that a 2.0 version would probably be needed after merging the current PRs is an even better argument for a 1.0 now. If it need to go to 4 or 5, even better. The more versions, the easier and more gradual are the steps to upgrade.

**Comment by iMacTia** [comment]
What I don't like is the idea of jumping into 1.0 now and jump again to 2.0 for what we now call 1.0.
That would just create a "useless" v1.0 which would see no minor version after it as we would jump straight to v2.0 (I don't want to release anything else after 0.14.0 and focus on 1.0).
I really don't see any advantage in doing so.

But if you prefer a more academic justification, here is what the [semantic versioning](http://semver.org) summary states very clearly:

> Given a version number MAJOR.MINOR.PATCH, increment the:
MAJOR version when you make incompatible API changes,
MINOR version when you add functionality in a backwards-compatible manner, and
PATCH version when you make backwards-compatible bug fixes.

Which is exactly what we did with Faraday 0.10 and above, and the reason for waiting till now for releasing v1.0

Moreover, down in the page:

> Major version zero (0.y.z) is for initial development. Anything may change at any time. The public API should not be considered stable.
Version 1.0.0 defines the public API. The way in which the version number is incremented after this release is dependent on this public API and how it changes.

Unfortunately, we're not there yet 😞. Different adapters have different behaviours and even common APIs (i.e. open/close timeouts) have different effects based on the adapter you use. The Errors/Exceptions you get back are inconsistent, the lack of documentation makes difficult implementing new adapters/middlewares. I could easily go on with the list and it was even worse when I started a year ago.

I personally believe these issues have to be addressed before we can really think about a v1.0 and that's what we're doing right now.

Faraday have had a huge success and I truly believe it's the most flexible HTTP client library out there, a sort of micro-framework that people love and trust, but I have to admit that even though it's used in production by many applications, it still lacks many things to be considered a mature gem.

I hope this (too) long post will help you understand why we took this decision and understand, even if not agree, with it.

**Comment by juanibiapina** [comment]
I stand by my concerns. None of your arguments are in favor of users of the library. Quoting the same source:

> If your software is being used in production, it should probably already be 1.0.0. If you have a stable API on which users have come to depend, you should be 1.0.0.

But let's end the discussion here. This should be enough for future reference.
